src/main/java/com/faforever/client/login/OAuthValuesReceiver.java

```diff
@@ -0,0 +1,147 @@
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketException;
+import java.net.URI;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/** Opens a minimal HTTP server that retrieves {@literal code} and {@literal state} from the browser. */
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+@RequiredArgsConstructor
+@Slf4j
+public class OAuthValuesReceiver {
+
+  private static final Pattern CODE_PATTERN = Pattern.compile("code=([^ &]+)");
+  private static final Pattern STATE_PATTERN = Pattern.compile("state=([^ &]+)");
+  private static final List<String> ALLOWED_HOSTS = List.of("localhost");
+
+  @VisibleForTesting
+  CompletableFuture<Integer> listenPort = new CompletableFuture<>();
+
+  private final ClientProperties clientProperties;
+  private final PlatformService platformService;
+  private final UserService userService;
+  private final I18n i18n;
+
+  public CompletableFuture<Values> receiveValues(URI fallbackRedirectUri, Optional<OAuthEndpoint> oAuthEndpoint) {
+    return CompletableFuture.supplyAsync(() -> {
+      Collection<URI> redirectUris = retrieveRedirectUris(fallbackRedirectUri, oAuthEndpoint);
+
+      for (URI uri : redirectUris) {
+        try {
+          return readWithUri(uri);
+        } catch (SocketException e) {
+          log.warn("Port " + uri.getPort() + " is probably already in use", e);
+        }
+      }
+      throw new IllegalStateException("Could not read from any redirect URI: " + redirectUris);
+    });
+  }
+
+  @SneakyThrows
+  private List<URI> retrieveRedirectUris(URI fallbackRedirectUri, Optional<OAuthEndpoint> oAuthEndpoint) {
+    JavaFxUtil.assertBackgroundThread();
+    if (clientProperties.isUseRemotePreferences()) {
+      OAuthEndpoint endpoint = oAuthEndpoint.orElseThrow(() -> new IllegalStateException("No endpoint has been specified"));
+      return endpoint.getRedirectUris().stream()
+          .filter(uri -> uri.getPort() > 0)
+          .filter(uri -> ALLOWED_HOSTS.contains(uri.getHost()))
+          .collect(Collectors.toList());
+    }
+    if (fallbackRedirectUri.toASCIIString().isBlank()) {
+      throw new NoFallbackRedirectUriException();
+    }
+    return Collections.singletonList(fallbackRedirectUri);
```

Comment:
We probably want to switch this so that the specified url takes priority over the remote preferences. Otherwise it makes testing on a local setup harder as it requires actively disabling the remote preferences check.

Reply:
How about now?

Reply:
Works